### ISSUE TITLE
decompTumor2Sig

### ISSUE THREAD

**Issue by rmpiro** [Bioconductor/Contributions#816]
Update the following URL to point to the GitHub repository of
the package you wish to submit to _Bioconductor_

- Repository: https://github.com/rmpiro/decompTumor2Sig

Confirm the following by editing each check box to '[x]'

- [X] I understand that by submitting my package to _Bioconductor_,
  the package source and all review commentary are visible to the
  general public.

- [X] I have read the _Bioconductor_ [Package Submission][2]
  instructions. My package is consistent with the _Bioconductor_
  [Package Guidelines][1].

[There is one notable exception here: preliminary versions of the package are already online at
http://rmpiro.net/decompTumor2Sig/index.html
Therefore, I hope that you can except a different version (1.3) from what you usually ask for (0.99);
please let me know if this is out of the question, or whether it is possible. Thank you!
Apart from this major issue, I have only some minor NOTEs, that I can address if necessary.]

- [X] I understand that a minimum requirement for package acceptance
  is to pass R CMD check and R CMD BiocCheck with no ERROR or WARNINGS.
  Passing these checks does not result in automatic acceptance. The
  package will then undergo a formal review and recommendations for
  acceptance regarding other Bioconductor standards will be addressed.

- [X] My package addresses statistical or bioinformatic issues related
  to the analysis and comprehension of high throughput genomic data.

- [X] I am committed to the long-term maintenance of my package. This
  includes monitoring the [support site][3] for issues that users may
  have, subscribing to the [bioc-devel][4] mailing list to stay aware
  of developments in the _Bioconductor_ community, responding promptly
  to requests for updates from the Core team in response to changes in
  _R_ or underlying software.

I am familiar with the essential aspects of _Bioconductor_ software
management, including:

- [X] The 'devel' branch for new packages and features.
- [X] The stable 'release' branch, made available every six
      months, for bug fixes.
- [X] _Bioconductor_ version control using [Git][5]
  (optionally [via GitHub][6]).

[Not an expert here, though :-)]

For help with submitting your package, please subscribe and post questions
to the [bioc-devel][4] mailing list.

[1]: https://bioconductor.org/developers/package-guidelines/
[2]: https://bioconductor.org/developers/package-submission/
[3]: https://support.bioconductor.org
[4]: https://stat.ethz.ch/mailman/listinfo/bioc-devel
[5]: http://bioconductor.org/developers/how-to/git/
[6]: http://bioconductor.org/developers/how-to/git/sync-existing-repositories/

**Comment by rmpiro** [comment]
There is one major problem with my submission: preliminary versions of the package are already online at
http://rmpiro.net/decompTumor2Sig/index.html
Therefore, I hope that you can except a different version (1.3) from what you usually ask for (0.99);
please let me know if this is out of the question, or whether it is possible. Thank you!
Apart from this major issue, I have only some minor NOTEs, that I can address if necessary.

**Comment by mtmorgan** [comment]
The code chunks in your vignette need to be evaluated, i.e., `{r}` after the triple tick.

**Comment by LiNk-NY** [comment]
Hi Rosario, @rmpiro 
Thank you for submitting your package to Bioconductor. 
Please see the review below.

Feel free to contact me here via issue comments.

Best regards,
Marcel 

---

# decompTumor2Sig #816

Please respond to Martin's @mtmorgan comments about specifying the language
in the vignette chunks for evaluation.

## DESCRIPTION

* It looks good. Consider using `Authors@R` instead of both the `Author` and
`Maintainer` field.
* Provide a `BugReports` URL field.

## NAMESPACE

* It looks good. It seems like `load*` may not be the best prefix to use for
such functions. It usually implies that you are putting something in the
`.GlobalEnv`.
* Many of your exports have long names. Make these shorter if possible.

## R 

* You may want to use `roxygen` for documentation being that you are readily
able to document your function in the same file.
* Use the native class checking mechanisms such as `is.numeric` instead of
`class(x) != "numeric"` or `is(X, "class")` for S4 classes.
* Use robust sequence creation `seq_len(N)` instead of `1:N`
(see also: seq_along).
* Provide a user-facing API for extracting data from `MutationFeatureData`
objects via accessor functions rather than using `getElement`.
* Avoid using `@` to access slots. There should be helper accessor functions
for extracting these data.
* Create a helper function for testing objects to be `data.frame`, `matrix`,
or `numeric`.
* Avoid the use of `sapply`

Minor:
* No need to test TRUE twice. `if (getElement(...) == TRUE)` can just be
`if (getElement(...))` unless the result of `getElement` is impure then use
`isTRUE`. 
* Use `vapply` to get an atomic vector from a `list` instead of
`unlist(lapply(...))`

## vignettes

* Use `BiocManager::install()` in the vignette instead of `biocLite`
* Make your code runnable by including `{r}` in the first code chunk delimiter.
* It is not recommended that you include a GitHub package dependency
(`pmsignature`)

**Comment by rmpiro** [comment]
Dear Martin, dear Marcel,
I replied by e-mail via the "reply-to" address, but this doesn't seem to have worked, as there is no trace of it here. So here's again an answer which addresses the observations you both have made:

Point-by-point reply (to Ramos, but this includes also the observation made by Martin):

DESCRIPTION
-----------

DONE It looks good. Consider using Authors@R instead of both the 
     Author and Maintainer field.

DONE Provide a BugReports URL field.


NAMESPACE
---------

DONE It looks good. It seems like load* may not be the best prefix to 
     use for such functions. It usually implies that you are putting 
     something in the .GlobalEnv.

  Reply: I've renamed all these functions in read*, which is more 
         appropriate.

DONE Many of your exports have long names. Make these shorter if 
     possible.

  Reply: I've shortened the longest ones. Are the names still too long?


R:
--

DONE You may want to use roxygen for documentation being that you are 
     readily able to document your function in the same file.

  Reply: I've changed the documentation, such that it can now be 
         generated and updated with roxygen2. Also the NAMESPACE is 
         now generated by roxygen2.
         For this purpose, I saved most of the functions into single 
         *R files (before that I had only two files: one for internal 
         functions and one for functions which are exported).

DONE Use the native class checking mechanisms such as is.numeric 
     instead of class(x) != "numeric" or is(X, "class") for S4 classes.

DONE Use robust sequence creation seq_len(N) instead of 1:N
     (see also: seq_along).

NOT DONE Provide a user-facing API for extracting data from 
         MutationFeatureData objects via accessor functions rather 
         than using getElement.

  Reply: MutationFeatureData is from the pmsignatures packages; The 
         authors of pmsignature themselves access the data using 
         "slot" within their code.
         Writing accessor functions for use in decomptumor2Sig would 
         require to integrate also the definitions of these classes, 
         which is potentially more (or at least equally) error prone 
         in case the classes change in future versions of pmsignature.
         But I understand that directly accessing their slots is 
         suboptimal because future changes of these classes may 
         significantly affect my code. I've therefore opted for 
         writing wrapper functions for accessing the slots, so that 
         within my main code the wrapper functions are used and
         these can more easily be modified in case the class 
         definitions change.
         Please also keep in mind that all of my functions which
         rely on pmsignature are _optional_ and my package can also
         be used without these functions.

DONE Avoid using @ to access slots. There should be helper accessor 
     functions for extracting these data.

   Reply: Done with one notable exception: the warpper functions I 
          just mentioned. Since MutationFeatureData and 
          EstimatedParameters of the pmsignature package don't come 
          with accessor function, I need to directly access the data 
          within my wrapper functions.

NOT DONE Create a helper function for testing objects to be 
         data.frame, matrix, or numeric.
     
  Reply: I'm not sure what you mean here. As suggested above, I've 
         replaced all class checks with the native class checking 
         mechanisms (e.g., is.matrix())
         The objects that decompTumor2Sig uses (signatures, genomes) 
         are basic object types (lists and numeric vectors and 
         matrices), so they don't require the definition of new helper 
         functions for class checking.

DONE Avoid the use of sapply

  Reply: I replaced all instances by vapply (a vector was the goal in 
         each case)


Minor:
------

NOT DONE No need to test TRUE twice. if (getElement(...) == TRUE) can 
         just be if (getElement(...)) unless the result of getElement 
         is impure then use isTRUE.

  Reply: That's correct, but in this case it is more human readable 
         because it is not immediately clear what kind of information 
         (here: boolean\logical) getElement(countData, 
         "transcriptionDirection") returns [I don't want to 
         need to think about this or check it next time I read the 
         code.]
         But if you think it's necessary to change this, I can do so.

DONE Use vapply to get an atomic vector from a list instead of
     unlist(lapply(...))

  Reply: (This required some other changes in mapSignatureSets() 
         because vapply unfortunately doesn't return a named vector 
         ...)


vignettes:
----------

NOT DONE Use BiocManager::install() in the vignette instead of biocLite

  Reply: I used biocLite only where the vignette explains how to 
         install this or other Bioconductor package. This is 
         equivalent to the installation instructions on the websites 
         of all Bioconductor packages, so I would prefer
         not to change it. Anyway, this code is not evaluated {r, 
         eval=FALSE}

DONE Make your code runnable by including {r} in the first code chunk 
     delimiter.

  Reply: The vignette code is now runnable; there are some exceptions:
  1) Some chunks that just show file formats etc. are, of course, not 
     run.
  2) The code chunks using the pmsignature package are not run because 
     the package is available only in GitHub (see below). However, all 
     these code chunks are tested and work perfectly fine once 
     pmsignature is installed. Actually, these code chunks were used 
     to create the corresponsing figures that I included with 
     include_graphics (knitr).

NOT DONE It is not recommended that you include a GitHub package 
         dependency (pmsignature)
         
  Reply: One of the major goals of the package is to do signature 
         refitting for both currently used models for mutational 
         signatures: the Alexandrov model (for which tools are already 
         available) and the Shiraishi model which is implemented by 
         pmsignature (and for which no other tools are available at
         the moment). It's unfortunate that the authors of pmsignature 
         never submitted their package to Bioconductor, but I think it 
         is anyway positive to have packages in Bioconductor that can 
         work with signatures taken from pmsignature (they might also 
         be obtained as flat files; pmsignature need not be 
         installed). It is important to illustrate how that works, so 
         I prefer to leave the code chunks in the vignette, but they 
         are not executed {r, eval=FALSE} so that users need not 
         install pmsignature (but these code chunks are all tested, 
         see above).
         As mentioned above, please not that all functions which
         directly rely on the pmsignature package are _optional_ and
         decompTumor2Sig can be used without! Our package only provides
         some extra functions to interface with pmsignature if the user
         wishes to do so.

Thanks a lot for the thorough review!

I hope my modifications and answers satisfy your concerns, otherwise
please let me know!

Best regards,
Rosario.

**Comment by LiNk-NY** [comment]
Hi Rosario, @rmpiro 

> NOT DONE Create a helper function for testing objects to be
data.frame, matrix, or numeric.

The point about creating a helper function for checking class is to avoid repetition
in the code and is optional. 
You have several instances that check for `!is.data.frame(x) & !is.matrix(x)` and
that could be made into a small helper function. 

> NOT DONE No need to test TRUE twice. if (getElement(...) == TRUE) can
just be if (getElement(...)) unless the result of getElement
is impure then use isTRUE.

It's a minor note but `getElement` shouldn't be used as a way to extract a `slot` or
`element` at this level in the code. The object should have an accessor function named
something like `isCondition(x)` so that it is truly readable (e.g., `if (isEven(...))`). 

> Reply: (This required some other changes in mapSignatureSets()
because vapply unfortunately doesn't return a named vector
...)

Note. `vapply` returns a named `vector` when a named `vector` is passed into it. 

```r
vapply(list(A = 1, B = 2), is.numeric, logical(1L))
   A    B 
TRUE TRUE 
```

> NOT DONE Use BiocManager::install() in the vignette instead of biocLite

We are transitioning all packages to use `BiocManager` instead of `BiocInstaller`. 
`BiocInstaller` will soon be deprecated. Your package will go into devel and it will
have to reflect the use of `BiocManager`. 

> NOT DONE It is not recommended that you include a GitHub package
dependency (pmsignature)

Unfortunately, we don't recommend the use of experimental `GitHub` packages in
vignette code as we want all our vignettes to have runnable code. I understand 
that this package provides useful functionality but its codebase may change at
any time and it can render your code stale.

Best regards,
Marcel

**Comment by rmpiro** [comment]
Hi Marcel,

I've updated to version 1.3.2.

Regarding the single points you raised:

1) I've created several helper functions for frequent class checks. Some of them were also interesting enough to be exported to the user (now also described in the manual and vignette; see NEWS for the list).

2) There's also a wrapper function for that; now it's named "haveTrDir" so that the name already indicates that a logical/Boolean value is returned; no need for a double check to ensure human readability.

3) I somehow didn't get names from the vapply, maybe because the input is a list, not a vector; anyway, it now works exactly as I wanted.

4) I wasn't aware that BiocInstaller was about to be replaced. I changed it to BiocManager in the vignette's description of the installation process.

5) I understand your point about having only runnable code in the vignette. But one of the major strengths of my package is that it can also work with the (in some ways more powerful) Shiraishi-type signatures which at the moment are provided only by pmsignature. My package can work without pmsignature (e.g. using Alexandrov-type signatures) but it's an additional strength to be able to work with both types of signatures. I somehow feel that it would be a pity to have these additional features but not describe them in the vignette. However, most of the vignette code is runnable without pmsignature. 

I hope this can still be acceptable because the paper about the package (which is currently under review) is mainly about Shiraishi signatures (this is the first tool that can work with them), so I can't (and don't want to) remove these additional features.
 
Thanks for all the effort you put into reviewing our stuff :-)

Best,
Rosario.

**Comment by LiNk-NY** [comment]
Hi Rosario, @rmpiro 

Thank you for your submission but as I've been advised, we cannot support software whose
long-term availability is uncertain. 

It would be best to have `pmsignature` prepared, submitted, and maintained
as part of Bioconductor. Then you can use this package as a dependency.
Please consider contacting the maintainer of that package. 

What we can do is add your annotation resources to `AnnotationHub`. 
Please see the vignette to see [how to add a resource to the hub]( http://www.bioconductor.org/packages/devel/bioc/vignettes/AnnotationHub/inst/doc/CreateAnAnnotationPackage.html).

Best regards,
Marcel

**Comment by rmpiro** [comment]
Hi Marcel, @LiNk-NY 

sorry for the late reply, I was on vacation.

I'm not so sure whether AnnotationHub would be the correct approach. It's not so much about having already predefined data from pmsignature that can't change anymore (no fixed data source). It's about accepting as input to my package also data/signatures that the user has previously computed using the pmsignature package.

Here's my plan:

I will completely decouple my code from the dependency of pmsignature doing the following:

1) I will implement own functions for visualizing the signatures which do not use functionalities provided by pmsignature.

2) Data from pmsignature which the user provides as input to my package's functions, will only be tested for compatibility/consistency with the pmsignature data format from pmsignature version 0.3.0, but my package will not use any functions/code of pmsignature for doing so.

This way the code of my package will be completely independent from the code of pmsignature and can 100% work without pmsignature ever being installed. It would still accept data from pmsignature in input in case the user does want to work with that.

Should the data format of pmsignature change, my package would simply complain that the input is not consistent with pmsignature version 0.3.0, but the code of my package would still perfectly run. In that case, I could update my package for newer versions of pmsignature, but my package would never "break" due to changes in pmsignature.

I will anyway contact the authors of pmsignature to see if they are interested in preparing it for Bioconductor (which would be nice, it's a nice package).

Would that be acceptable?

I've nearly completed point 2 (check for consistency of input data) and have a solution for part of point 1 (own graphical output for one of the two signature types my package supports; need to implement the second).

I think I can submit a new version within a few days.

Since I also have to submit the final version of my paper in these days (where I would like to specify in case my package will appear on Bioconductor), please let me know whether my proposal is generally acceptable or not.

To summarize the proposal: my package will accept data produced by pmsignature version 0.3.0 as input, but it will be completely independent of any code of pmsignature (including the vignette which will be fully runnable), so that my package will run smoothly also in case a new version of pmsignature is published (in that case it simply won't accept the data from the new version if it's not compatible to 0.3.0, or the user can convert the data him-/herself, which is actually not so difficult using pmsignature itself).

Please let me know asap, and thanks a lot for the effort you put into this :-)

Best regards,
Rosario.

**Comment by LiNk-NY** [comment]
Hi Rosario, @rmpiro 
Yes, this looks like a viable solution. 

Having your code independent of `pmsignature` would be best since you
can't guarantee that `pmsignature` will be stable. 

It's also a good idea to nudge the authors of `pmsignature` into preparing their
package for Bioconductor. We'd be happy to review it. 

Thanks! 
-MR

**Comment by rmpiro** [comment]
Hi Marcel, @LiNk-NY 

Ok, thanks for the quick reply. Is it ok for you then, if I write in the paper that decompTumor2Sig will (sooner or later) be integrated in Bioconductor? If I completely decouble the code from pmsignature (while still accepting their data as input), we should be at a good point, right?

Thanks again,
Rosario.

**Comment by LiNk-NY** [comment]
Hi Rosario, @rmpiro 
Yes, I would first remove `pmsignature` specific code from the package and push your changes along with a version bump so that I can have a look at it. 
Thanks!

**Comment by rmpiro** [comment]
Hi Marcel, @LiNk-NY 

that was a major effort (especially the plot functions), but I removed all code dependency of the pmsignature package. I've implemented completely new plot functions now. The functions that allow to accept input from pmsignature remain, but they do not use any code of pmsignature, they only verify that the input data is acceptable to be parsed for decompTumor2Sig. Also, please remember that input from pmsignature is not necessary and there are different ways of getting input for my package (reading VCF and MPF files, converting from VRanges objects, accessing signatures from the COSMIC website or flat files, etc.)

I hope this looks better now :-)

Thanks again for the effort!

**Comment by LiNk-NY** [comment]
Hi Rosario, @rmpiro 
Can you push the changes and bump your package version? 
Thanks!

**Comment by LiNk-NY** [comment]
Hi Rosario, @rmpiro 
Please merge your changes to your master branch on GitHub and bump your version so that the Single Package Builder (SPB) can rebuild and test your package. 
```
git checkout master
git merge <name of branch here>
# bump 'z' version in your package
git push origin master
```
Thanks.

**Comment by rmpiro** [comment]
Hi Marcel, @LiNk-NY 
I hope I did it correctly this time. Thanks for the patience.
Ciao,
Rosario.

**Comment by lshep** [comment]
As there has been no subsequent builds -  I would assume the webhook is not initialized - Please see the instructions https://github.com/Bioconductor/Contributions#adding-a-web-hook    Once the webhook is initilized please try another version bump

**Comment by LiNk-NY** [comment]
Thanks Lori!

**Comment by rmpiro** [comment]
Hi Marcel @LiNk-NY , hi Lori @lshep ,

sorry for the long silence, I've been on a pilgrimage.

I hope I finally managed to submit everything correctly. Thanks for your help!

Ciao,
Rosario.

**Comment by lshep** [comment]
Please fix all ERRORs before @LiNk-NY does a technical review -  It still seems like you have pmsignature as a Suggest in the DESCRIPTION file -  Bioconductor will not allow packages that are not in CRAN or Bioconductor to be listed in the DESCRIPTION file.

**Comment by LiNk-NY** [comment]
Hi Rosario, @rmpiro 
Can you please provide some updates? 
Thanks,
Marcel

**Comment by rmpiro** [comment]
Hi Marcel @LiNk-NY , hi Ishep @lshep ,

sorry for the long silence. I was first out of office for a while (conference and vacation) and then sick for an entire week. I'm finally back "online" and have just submitted an updated version.

There will be at least one error remaining: the error due to the package which I would prefer not to lower to 0.99 for the reasons outlined above. I hope the other issues are solved ... let's see :-)

**Comment by LiNk-NY** [comment]
Hi Rosario, @rmpiro 
Can you try setting the version to `1.99.0`?
This is probably better since your package will have another life on Bioconductor.

Best regards,
Marcel

**Comment by rmpiro** [comment]
Hi Marcel @LiNk-NY ,

thanks for the suggestion to set the version to 1.99.0. The error remains, but I guess it's anyway a good idea to have the version significantly different (and move to 2.y.z) when the package will be included in Bioconductor.

I hope that appart from the error due to the version number everything else is ok (at least the rest seems to look fine for me).

Best,
Rosario.

**Comment by LiNk-NY** [comment]
Lori, @lshep, is there any way to avoid this error? 
Thanks.

**Comment by lshep** [comment]
No just ignore it.

**Comment by rmpiro** [comment]
Hi Marcel @LiNk-NY , hi Lori @lshep ,
great! Thanks a lot for accepting the package :-)
Should I move the version number to 2.0.0 now?

**Comment by lshep** [comment]
No - It will remain 1.99  in devel and when it is released in spring it will automatically get bumped to 2.0

**Comment by rmpiro** [comment]
@lshep : Ok, thanks for the clarification :-)

**Comment by mtmorgan** [comment]
The master branch of your GitHub repository has been added to Bioconductor's git repository.

To use the git.bioconductor.org repository, we need an 'ssh' key to associate with your github user name. If your GitHub account already has ssh public keys (https://github.com/rmpiro.keys is not empty), then no further steps are required. Otherwise, do the following:

1. [Add an SSH key to your github account][]
1. [Submit your SSH key to Bioconductor][]

See further instructions at

https://bioconductor.org/developers/how-to/git/

for working with this repository. See especially

https://bioconductor.org/developers/how-to/git/new-package-workflow/
https://bioconductor.org/developers/how-to/git/sync-existing-repositories/

to keep your GitHub and Bioconductor repositories in sync.

Your package will be included in the next nigthly 'devel' build (check-out from git at about 6 pm Eastern; build completion around 2pm Eastern the next day) at

https://bioconductor.org/checkResults/

(Builds sometimes fail, so ensure that the date stamps on the main landing page are consistent with the addition of your package). Once the package builds successfully, you package will be available for download in the 'Devel' version of Bioconductor using `BiocManager::install("decompTumor2Sig")`. The package 'landing page' will be created at

https://bioconductor.org/packages/decompTumor2Sig

If you have any questions, please contact the bioc-devel mailing list (https://stat.ethz.ch/mailman/listinfo/bioc-devel); this issue will not be monitored further.

[Add an SSH key to your github account]: https://help.github.com/articles/connecting-to-github-with-ssh/
[Submit your SSH key to Bioconductor]: https://git.bioconductor.org/BiocCredentials/
